py/desispec/io/emlinefit.py

```diff
@@ -0,0 +1,567 @@
+        Use the FIBERMAP extension.
+        Do not use desitarget.targets.main_cmx_or_sv, as we only read TARGETID and the mask column,
+            so it is faster; but we use same philosophy.
+    """
+    # AR get the *DESI_TARGET colum
+    dtkeys = [
+        key for key in fits.open(redrockfn)["FIBERMAP"].columns.names
+            if key in ["CMX_TARGET", "SV1_DESI_TARGET", "SV2_DESI_TARGET", "SV3_DESI_TARGET"]
+    ]
+    # AR if no match, then assume it is a main catalog
+    if len(dtkeys) == 0:
+        dtkey = "DESI_TARGET"
+        from desitarget.targetmask import desi_mask as mask
+    elif len(dtkeys) == 1:
+        dtkey = dtkeys[0]
+        if dtkey == "CMX_TARGET":
+            from desitarget.cmx.cmx_targetmask import cmx_mask as mask
+        if dtkey == "SV1_DESI_TARGET":
+            from desitarget.sv1.sv1_targetmask import desi_mask as mask
+        if dtkey == "SV2_DESI_TARGET":
+            from desitarget.sv2.sv2_targetmask import desi_mask as mask
+        if dtkey == "SV3_DESI_TARGET":
+            from desitarget.sv3.sv3_targetmask import desi_mask as mask
+    else:
+        log.info(
+            "found {}>1 matching keys: {}; 0 or 1 key expected; exiting".format(
+                len(dtkeys), dtkeys
+            )
+        )
+        sys.exit(1)
+    allowed_bitnames = mask.names()
+    # AR read + select the targetids
+    d = fitsio.read(redrockfn, columns=["TARGETID", dtkey], ext="FIBERMAP")
+    sel = np.zeros(len(d), dtype=bool)
+    for bitname in bitnames.split(","):
+        if bitname not in allowed_bitnames:
+            log.info(
+                "{} not in allowed bitnames for {} ({}; exiting)".format(
+                    bitname, dtkey, allowed_bitnames
+                )
+            )
+            sys.exit(1)
+        sel |= (d[dtkey] & mask[bitname]) > 0
+    targetids = d["TARGETID"][sel]
+    log.info("selecting {} targets with {} in {}".format(sel.sum(), bitnames, dtkey))
+    return targetids
+
+
+def get_rf_em_waves(emname):
+    """
+    Returns the rest-frame, vacuum, wavelengths.
+    Args:
+        emname, from: "OII", "HDELTA", "HGAMMA", "HBETA", "OIII", "HALPHA" (string)
+
+    Returns:
+        rf_em_waves: rest-frame wavelength(s) (np array)
+
+    Notes:
+        For OII and OIII returns a two-elements array; one-element array otherwise.
+        https://github.com/desihub/fastspecfit/blob/60393296e0cc466858f70a5d021d02693eff9375/py/fastspecfit/data/emlines.ecsv
+    """
+    if emname == "OII":
+        rf_em_waves = np.array([3727.092, 3729.874])
+    if emname == "OIII":
+        rf_em_waves = np.array([4960.295, 5008.239])
+    if emname == "HALPHA":
+        rf_em_waves = np.array([6564.613])
+    if emname == "HBETA":
+        rf_em_waves = np.array([4862.683])
+    if emname == "HGAMMA":
+        rf_em_waves = np.array([4341.684])
+    if emname == "HDELTA":
+        rf_em_waves = np.array([4102.892])
+    return rf_em_waves
+
+
+def emlines_gaussfit(
```

Comment:
This function is the core algorithm that doesn't do I/O (good), but should be in a separate non-I/O module like `desispec.emlinefit`, leaving `desispec.io.emlinefit` for functions that are very specific to I/O itself.  `get_rf_em_waves` would also go in that module.

Reply:
done.